### PR TITLE
[FLINK-27546] Add append only writer which implements the RecordWriter interface

### PR DESCRIPTION
**Pull request by openinx**
It's a sub PR for https://github.com/apache/flink-table-store/pull/99, which is trying to add an AppendOnlyWriter and unit test to accept records without any key or sort order.

### CODE REVIEW

File: flink-table-store-core/src/main/java/org/apache/flink/table/store/file/writer/AppendOnlyWriter.java [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+    }
+
+    @Override
+    public List<DataFileMeta> close() throws Exception {
+        sync();
+
+        List<DataFileMeta> result = Lists.newArrayList();
+        if (writer != null) {
+            // Abort this writer to clear uncommitted files.
+            writer.abort();
```
Comment (openinx) [inline flink-table-store-core/src/main/java/org/apache/flink/table/store/file/writer/AppendOnlyWriter.java:114]: I'm just curious that the `RecordWriter#close` will close the writer and **delete the uncommitted files** automatically,  which is a bit mismatching the normal `close` semantic of writer.  [1]

Actually, the behavior is more matching the `abort` semantic. So I filed a separate issue to address this issue. I mean we can define separate `abort()` and `close()` methods in this interface,  and the `close()` will just close the opening writer, and the `abort()` will clean the uncommitted files. I think its semantic is more clear.

[1] https://github.com/apache/flink-table-store/blob/1fedb6586ba292bedf00bb1b6922d5110a2f7fa8/flink-table-store-core/src/main/java/org/apache/flink/table/store/file/writer/RecordWriter.java#L51

Comment (openinx) [reply flink-table-store-core/src/main/java/org/apache/flink/table/store/file/writer/AppendOnlyWriter.java:111]: Filed a separate issue for this: https://issues.apache.org/jira/browse/FLINK-27553

Comment (openinx) [reply flink-table-store-core/src/main/java/org/apache/flink/table/store/file/writer/AppendOnlyWriter.java:111]: Filed a separate issue for this: https://issues.apache.org/jira/browse/FLINK-27553

File: flink-table-store-core/src/test/java/org/apache/flink/table/store/file/writer/TestAppendOnlyWriter.java [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.io.TempDir;
+
+import java.util.Collections;
+import java.util.List;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+/** Test the correctness for {@link AppendOnlyWriter}. */
+public class TestAppendOnlyWriter {
```
Comment (JingsongLi) [inline flink-table-store-core/src/test/java/org/apache/flink/table/store/file/writer/TestAppendOnlyWriter.java:49]: Flink style is `AppendOnlyWriterTest`.
Only the classes `XXXTest` and `XXXITCase` will be run.

File: flink-table-store-core/src/main/java/org/apache/flink/table/store/file/writer/AppendOnlyWriter.java [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+
+    private class RowRollingWriter extends RollingFileWriter<RowData, DataFileMeta> {
+
+        public RowRollingWriter(Supplier<RowFileWriter> writerFactory, long targetFileSize) {
+            super(writerFactory, targetFileSize);
+        }
+    }
+
+    private class RowFileWriter extends BaseFileWriter<RowData, DataFileMeta> {
+        private final long startSeqNum;
```
Comment (JingsongLi) [inline flink-table-store-core/src/main/java/org/apache/flink/table/store/file/writer/AppendOnlyWriter.java:133]: nit: `minSequenceNumber`?

File: flink-table-store-core/src/main/java/org/apache/flink/table/store/file/writer/AppendOnlyWriter.java (hunk 1 above)
Comment (JingsongLi) [reply flink-table-store-core/src/main/java/org/apache/flink/table/store/file/writer/AppendOnlyWriter.java:111]: Sounds good to me~

File: flink-table-store-core/src/main/java/org/apache/flink/table/store/file/writer/AppendOnlyWriter.java [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+    @Override
+    public void sync() throws Exception {
+        // Do nothing here, as this writer don't introduce any async compaction thread currently.
+    }
+
+    @Override
+    public List<DataFileMeta> close() throws Exception {
+        sync();
+
+        List<DataFileMeta> result = Lists.newArrayList();
```
Comment (JingsongLi) [inline flink-table-store-core/src/main/java/org/apache/flink/table/store/file/writer/AppendOnlyWriter.java:108]: NIT: `new ArrayList<>();`
We don't use guava if we don't have to.

File: flink-table-store-core/src/main/java/org/apache/flink/table/store/file/writer/AppendOnlyWriter.java [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+    @Override
+    public Increment prepareCommit() throws Exception {
+        List<DataFileMeta> newFiles = Lists.newArrayList();
+
+        if (writer != null) {
+            writer.close();
+            newFiles.addAll(writer.result());
+
+            // Reopen the writer to accept further records.
+            writer = createRollingRowWriter();
```
Comment (JingsongLi) [inline flink-table-store-core/src/main/java/org/apache/flink/table/store/file/writer/AppendOnlyWriter.java:93]: Can we create this writer lazily? I can see always have a empty file in batch insert.

File: flink-table-store-core/src/main/java/org/apache/flink/table/store/file/data/DataFileMeta.java [hunk 6]
```diff
@@ -49,6 +57,26 @@ public class DataFileMeta {
     private final long maxSequenceNumber;
     private final int level;
 
+    public DataFileMeta(
```
Comment (JingsongLi) [inline flink-table-store-core/src/main/java/org/apache/flink/table/store/file/data/DataFileMeta.java:60]: Can we have a static method , maybe just `DataFileMeta.fromAppend`

Comment (openinx) [reply flink-table-store-core/src/main/java/org/apache/flink/table/store/file/data/DataFileMeta.java:60]: Okay, that seems to be more clear.

File: flink-table-store-core/src/main/java/org/apache/flink/table/store/file/writer/AppendOnlyWriter.java (hunk 5 above)
Comment (openinx) [reply flink-table-store-core/src/main/java/org/apache/flink/table/store/file/writer/AppendOnlyWriter.java:93]: Thanks for the careful reviewing.  In fact, here we are creating the `RollingFileWriter` which will initializing the `DataFileWriter` lazily ( The `DataFileWriter` is the real writer that will try to open file descriptor to write records ). 

Please check this lines: https://github.com/apache/flink-table-store/blob/master/flink-table-store-core/src/main/java/org/apache/flink/table/store/file/writer/RollingFileWriter.java#L58-L69

So in theory,  we won't add any extra empty `DataFileMeta` into the table format, even if we open an `AppendOnlyWriter` and call `prepareCommit` without writing any record.

File: flink-table-store-core/src/main/java/org/apache/flink/table/store/file/writer/AppendOnlyWriter.java (hunk 4 above)
Comment (openinx) [reply flink-table-store-core/src/main/java/org/apache/flink/table/store/file/writer/AppendOnlyWriter.java:108]: Okay, let's follow the coding rule.

File: flink-table-store-core/src/test/java/org/apache/flink/table/store/file/writer/TestAppendOnlyWriter.java (hunk 2 above)
Comment (openinx) [reply flink-table-store-core/src/test/java/org/apache/flink/table/store/file/writer/TestAppendOnlyWriter.java:49]: Thanks for the remaining,  let me rename this class name.

**Comment by openinx** [comment]
The failure e2e case is: 

```
Error:  testWithoutPk  Time elapsed: 88.458 s  <<< ERROR!
org.testcontainers.containers.ContainerLaunchException: Container startup failed
	at org.testcontainers.containers.GenericContainer.doStart(GenericContainer.java:336)
	at org.testcontainers.containers.GenericContainer.start(GenericContainer.java:317)
	at org.testcontainers.containers.ContainerisedDockerCompose.invoke(DockerComposeContainer.java:646)
	at org.testcontainers.containers.DockerComposeContainer.runWithCompose(DockerComposeContainer.java:309)
	at org.testcontainers.containers.DockerComposeContainer.createServices(DockerComposeContainer.java:233)
	at org.testcontainers.containers.DockerComposeContainer.start(DockerComposeContainer.java:177)
	at org.apache.flink.table.store.tests.E2eTestBase.before(E2eTestBase.java:102)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.platform.commons.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:725)
	at org.junit.jupiter.engine.execution.MethodInvocation.proceed(MethodInvocation.java:60)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$ValidatingInvocation.proceed(InvocationInterceptorChain.java:131)
	at org.junit.jupiter.engine.extension.TimeoutExtension.intercept(TimeoutExtension.java:149)
	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptLifecycleMethod(TimeoutExtension.java:126)
	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptBeforeEachMethod(TimeoutExtension.java:76)
	at org.junit.jupiter.engine.execution.ExecutableInvoker$ReflectiveInterceptorCall.lambda$ofVoidMethod$0(ExecutableInvoker.java:115)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.lambda$invoke$0(ExecutableInvoker.java:105)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$InterceptedInvocation.proceed(InvocationInterceptorChain.java:106)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.proceed(InvocationInterceptorChain.java:64)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.chainAndInvoke(InvocationInterceptorChain.java:45)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.invoke(InvocationInterceptorChain.java:37)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:104)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:98)
	at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.invokeMethodInExtensionContext(ClassBasedTestDescriptor.java:506)
	at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.lambda$synthesizeBeforeEachMethodAdapter$21(ClassBasedTestDescriptor.java:491)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$invokeBeforeEachMethods$3(TestMethodTestDescriptor.java:171)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$invokeBeforeMethodsOrCallbacksUntilExceptionOccurs$6(TestMethodTestDescriptor.java:199)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.invokeBeforeMethodsOrCallbacksUntilExceptionOccurs(TestMethodTestDescriptor.java:199)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.invokeBeforeEachMethods(TestMethodTestDescriptor.java:168)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:131)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:66)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:151)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
	at java.util.ArrayList.forEach(ArrayList.java:1259)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:41)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:155)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
	at java.util.ArrayList.forEach(ArrayList.java:1259)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:41)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:155)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.submit(SameThreadHierarchicalTestExecutorService.java:35)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor.execute(HierarchicalTestExecutor.java:57)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestEngine.execute(HierarchicalTestEngine.java:54)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:220)
	at org.junit.platform.launcher.core.DefaultLauncher.lambda$execute$6(DefaultLauncher.java:188)
	at org.junit.platform.launcher.core.DefaultLauncher.withInterceptedStreams(DefaultLauncher.java:202)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:181)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:128)
	at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.invokeAllTests(JUnitPlatformProvider.java:142)
	at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.invoke(JUnitPlatformProvider.java:109)
	at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:384)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:345)
	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:126)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:418)
Caused by: org.rnorth.ducttape.RetryCountExceededException: Retry limit hit with exception
	at org.rnorth.ducttape.unreliables.Unreliables.retryUntilSuccess(Unreliables.java:88)
	at org.testcontainers.containers.GenericContainer.doStart(GenericContainer.java:[329](https://github.com/apache/flink-table-store/runs/6368915717?check_suite_focus=true#step:5:329))
	... 75 more
Caused by: org.testcontainers.containers.ContainerLaunchException: Could not create/start container
	at org.testcontainers.containers.GenericContainer.tryStart(GenericContainer.java:525)
	at org.testcontainers.containers.GenericContainer.lambda$doStart$0(GenericContainer.java:[331](https://github.com/apache/flink-table-store/runs/6368915717?check_suite_focus=true#step:5:331))
	at org.rnorth.ducttape.unreliables.Unreliables.retryUntilSuccess(Unreliables.java:81)
	... 76 more
Caused by: java.lang.IllegalStateException: Container did not start correctly.
	at org.testcontainers.containers.GenericContainer.tryStart(GenericContainer.java:463)
	... 78 more
```

Let's just reopen to rerun the unit test.

File: flink-table-store-core/src/main/java/org/apache/flink/table/store/file/writer/AppendOnlyWriter.java (hunk 5 above)
Comment (openinx) [reply flink-table-store-core/src/main/java/org/apache/flink/table/store/file/writer/AppendOnlyWriter.java:93]: Added an unit test for this: https://github.com/apache/flink-table-store/pull/115/files#diff-d75bc35d7bf343c2eb454f30bc548fa52f4268d6f2147d30537b2332b5176c9aR68-R80

File: flink-table-store-core/src/main/java/org/apache/flink/table/store/file/writer/AppendOnlyWriter.java [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+ * A {@link RecordWriter} implementation that only accepts records which are always insert
+ * operations and don't have any unique keys or sort keys.
+ */
+public class AppendOnlyWriter implements RecordWriter {
+    private final BulkWriter.Factory<RowData> writerFactory;
+    private final RowType writeSchema;
+    private final long targetFileSize;
+    private final DataFilePathFactory pathFactory;
+    private final FileStatsExtractor fileStatsExtractor;
+    private final AtomicLong nextSeqNum;
```
Comment (tsreaper) [inline flink-table-store-core/src/main/java/org/apache/flink/table/store/file/writer/AppendOnlyWriter.java:53]: [Previous discussion](https://github.com/apache/flink-table-store/pull/99#discussion_r868746358)

Inner class does not need final members. A primitive `long` is enough. Try it.

Comment (openinx) [reply flink-table-store-core/src/main/java/org/apache/flink/table/store/file/writer/AppendOnlyWriter.java]: Yes, you are right.  Lambda implementation needs final modifier, while inner class don't. https://stackoverflow.com/questions/34865383/variable-used-in-lambda-expression-should-be-final-or-effectively-final

**Review by tsreaper — APPROVED** [review]

File: flink-table-store-core/src/main/java/org/apache/flink/table/store/file/data/DataFileMeta.java [hunk 8]
```diff
@@ -49,6 +57,46 @@ public class DataFileMeta {
...
+                EMPTY_MIN_KEY,
+                EMPTY_MAX_KEY,
+                EMPTY_KEY_STATS,
+                rowStats,
+                minSequenceNumber,
+                maxSequenceNumber,
+                DUMMY_LEVEL);
+    }
+
+    public DataFileMeta(
```
Comment (JingsongLi) [inline flink-table-store-core/src/main/java/org/apache/flink/table/store/file/data/DataFileMeta.java:80]: Remove this method?

File: flink-table-store-core/src/main/java/org/apache/flink/table/store/file/mergetree/Increment.java [hunk 9]
```diff
@@ -35,12 +35,19 @@
...
+    private static final List<DataFileMeta> EMPTY_COMPACT_BEFORE = Collections.emptyList();
+    private static final List<DataFileMeta> EMPTY_COMPACT_AFTER = Collections.emptyList();
+
     private final List<DataFileMeta> newFiles;
 
     private final List<DataFileMeta> compactBefore;
 
     private final List<DataFileMeta> compactAfter;
 
+    public Increment(List<DataFileMeta> newFiles) {
```
Comment (JingsongLi) [inline flink-table-store-core/src/main/java/org/apache/flink/table/store/file/mergetree/Increment.java:47]: Ditto: introduce a `forAppend`

File: flink-table-store-core/src/main/java/org/apache/flink/table/store/file/writer/AppendOnlyWriter.java [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+        this.nextSeqNum = maxWroteSeqNumber + 1;
+
+        this.writer = createRollingRowWriter();
+    }
+
+    @Override
+    public void write(ValueKind valueKind, RowData key, RowData value) throws Exception {
+        Preconditions.checkArgument(
+                valueKind == ValueKind.ADD,
+                "Append-only writer cannot accept ValueKind: " + valueKind);
```
Comment (JingsongLi) [inline flink-table-store-core/src/main/java/org/apache/flink/table/store/file/writer/AppendOnlyWriter.java:77]: We should avoid string concat here. Use 
```
checkArgument(
            boolean condition,
            @Nullable String errorMessageTemplate,
            @Nullable Object... errorMessageArgs
```

**Review by JingsongLi — COMMENTED** [review]
Thanks for the update, left some comments.

**Review by JingsongLi — APPROVED** [review]
Looks good to me!

**Pull request merged**